### ISSUE TITLE
Proposal and RFC: Data modelling (ish)

### ISSUE THREAD

**Issue by alistairjcbrown** [apoclyps/my-dev-space#249]
**Caveat** - this is mostly a brain dump of thoughts, please feel free to comment, pick apart, critique etc. as much as you can / want to!

Also, this isn't a plan for "we must do this right now", more a starting point for a conversation on how all this data fits together. My main thought is we should probably get the data structures consistent (what data we hold about each thing, how we name it, etc.) and then we can figure out the relationships between everything.

## Data Modelling

Things we're to model
 - Events
 - Videos
 - Meetups
 - Users

![untitled drawing](https://user-images.githubusercontent.com/635903/42738821-592ec7e8-8882-11e8-8579-8cd11da8aaf2.png)

### Event
```json
{
  "id": "uuid (required)",
  "name": "Name of the event (required)",
  "description": "Description of the event (required)",
  "topics": "Topics being discussed at the event (optional)", 
  "time": "Date and time of event start (required)",
  "duration": "Duration of the event in seconds (optional)",
  "entry": "Public, Registration, Ticket, etc.",
  "url": "Url to original source (required)",
  "image_url": "Url of image for the event (optional)",
  "meetup": "Meetup, eg. BelfastJS - Foreign key? (required)",
  "source": "Event source, eg. Meetup.com - Foreign key? (required)",
  "created_at": "Created at date and time, generated by database (system)",
  "updated_at": "Updated at date and time, generated by database (system)",
  "deleted_at": "Deleted at date and time, used to soft deleted (system)"
}
```
❓ Will an event only be organised by a single meetup?
❓ Is is possible we'll need to deduplicate events from multiple sources?
❓ Will events ever be soft deleted and updated, or just resynced with the origin source?
❓ Should call for speakers be part of an event, or at the meetup level?
❓ Should "speakers" and "users" differ?

### User
```json
{
  "id": "uuid (required)",
  "name": "Users name, however they want it formatted (required)",
  "avatar": "Url of image representing user (optional)",
  "bio": "Information about the user (optional)",
  "contact": "Some form of contact; email, twitter, github, etc. (optional)",
  "topics": "Topics the user is interested in / would talk about (optional)", 
  "diversification": "Things the user feels would provide a differing viewpoint to the mainstream, eg. member of underrepresented group (optional)", 
  "location": "Main location the user is present in (attending meetups, speaking, etc.) (optional)",
  "created_at": "Created at date and time, generated by database (system)",
  "updated_at": "Updated at date and time, generated by database (system)",
  "deleted_at": "Deleted at date and time, used to soft deleted (system)"
}
```
❓ Is this linked to an authenticated account which can update these details?
❓ Is a source necessary for a speaker? This feels like something we'll only want to initially sync?
❓ Should topics be pushed out into their own table for easier grouping?
❓ How freeform and granular should location be? City? Country? Vague ~50 mile geographic area?

### Video
```json
{
  "id": "uuid (required)",
  "name": "Title of the video (required)",
  "description": "Description of the video content (required)",
  "topics": "Topics being discussed in the video (optional)", 
  "duration": "Duration of the video in seconds (required)",
  "url": "Url to original source (required)",
  "image_url": "Url of image for the video (optional)",
  "meetup": "Meetup, eg. BelfastJS - Foreign key? (required)",
  "source": "Video source, eg. YouTube.com - Foreign key? (required)",
  "created_at": "Created at date and time, generated by database (system)",
  "updated_at": "Updated at date and time, generated by database (system)",
  "deleted_at": "Deleted at date and time, used to soft deleted (system)"
}
```
❓ Is a channel necessary for a video? Should the "channel" be something we model as a "meetup"?
 - Does that make sense to model something like NIDevConf as a ticketed meetup? 

### Meetup
```json
{
  "id": "uuid (required)",
  "name": "Name of the meetup (required)",
  "description": "Description of the meetup (required)",
  "topics": "Topics generally being discussed at the meetup (optional)", 
  "url": "Url to original source (required)",
  "image_url": "Url of image for the event (optional)",
  "source": "Event source, eg. Meetup.com - Foreign key? (required)",
  "created_at": "Created at date and time, generated by database (system)",
  "updated_at": "Updated at date and time, generated by database (system)",
  "deleted_at": "Deleted at date and time, used to soft deleted (system)"
}
```
❓ Maybe this is where we could set some defaults for event durations, event entry requirements, etc.?

---

## General thoughts

 - I'm not a fan of the changes I made the the events endpoint - I'd love it if the events endpoint just took a date range, and then we could just ask for events from 12 hours ago until 1 year from now and present those (perhaps paginated)
 - Not something to enforce, but a list of topics, from both meetups, events and users, would be an interesting way to start to pair people and meetups together. Having historical data on who's spoken at which meetups would also be nice, but I imagine that'll be pretty difficult to generate without manual intervention.
 - If we start to take user generated input (user accounts, meetups not on other sites, topics, etc.) should we have some separation from that data and synced data so that, if required, we can resync data from remote sources, but keep user generated content? Or is that overkill?

**Comment by apoclyps** [comment]
Thanks for raising this; It's something that really needs to be corrected before promoting the current standing build to production. I took a crack at refactoring the existing models into 3NF. The following tables were produced as a result of taking onboard the feedback provided. I'm open to changing any of this but I hope it addresses most of the above.

```
Schema |             Name              | Type  |  Owner
--------+-------------------------------+-------+----------
public | channel                       | table | postgres
public | channel_topic_association     | table | postgres
public | diversity                     | table | postgres
public | entry                         | table | postgres
public | event                         | table | postgres
public | event_entry_association       | table | postgres
public | event_topic_association       | table | postgres
public | meetup                        | table | postgres
public | meetup_event_association      | table | postgres
public | speaker                       | table | postgres
public | speaker_diversity_association | table | postgres
public | speaker_topic_association     | table | postgres
public | topic                         | table | postgres
public | video                         | table | postgres
public | video_topic_association       | table | postgres
```

For further details on the proposed schema; the following [PR](https://github.com/apoclyps/my-dev-space/pull/253) details all the changes but the majority of the model changes can be found [here ](https://github.com/apoclyps/my-dev-space/blob/13a1e2396ed47c5e02c341f77ca87a606a79049c/services/events/project/api/models.py)

**ERD Diagram**
Relationships between tables are highlightened however the fields have not yet been added
![untitled diagram](https://user-images.githubusercontent.com/1443700/42787143-1ff8c2e2-8951-11e8-92c1-a11c6f807a8f.jpg)

**Comment by apoclyps** [comment]
The first pass at the revised schema has been applied to staging from #253 : http://my-dev-space-staging-alb-2128504978.us-east-1.elb.amazonaws.com/event

**Comment by apoclyps** [comment]
Closed by #253
